### PR TITLE
Add StableLM-3B 4E1T to Keras Hub

### PR DESCRIPTION
**Pull request by Bond099**
This PR adds the StableLM-3B 4E1T model to Keras Hub. However, numerical matching with the Hugging Face implementation is still in progress.

### CODE REVIEW

**Comment by Bond099** [comment]
Here is a comparison of numerics with Hugging Face in [Colab](https://colab.research.google.com/drive/17hpMiC2vRP783ss1vqq3lCoo50qY1NRi?usp=sharing). It works fine with an absolute tolerance of 1e-3, but not with 1e-5. Please guide.

File: keras_hub/src/models/stablelm/stablelm_attention.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    """StableLMAttention layer.
+
+    This layer implements the attention mechanism for StableLM-3B4E1T, featuring
+    multi-head self-attention with partial rotary position embeddings applied
+    to a fraction of the head dimensions, as specified by `rotary_percentage`.
+    It is adapted from the LlamaAttention layer with modifications to align
+    with StableLM's official configuration(https://github.com/Stability-AI/StableLM/blob/main/configs/stablelm-3b-4e1t.yml).
+
+    Args:
+        num_query_heads (int): Number of attention heads for queries.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/stablelm/stablelm_attention.py:20]: format type hints similar to other models

File: keras_hub/src/models/stablelm/stablelm_backbone.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    as described in the official documentation. It is a decoder-only transformer similar
+    to LLaMA with modifications including partial rotary position embeddings and
+    LayerNorm with learned bias terms. It includes the embedding lookups and transformer
+    layers.
+
+    The default constructor provides a fully customizable, randomly initialized
+    StableLM model with any number of layers, heads, and embedding dimensions.
+
+    Args:
+        vocabulary_size (int): The size of the token vocabulary.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/stablelm/stablelm_backbone.py:29]: fix type hint format

File: keras_hub/src/models/stablelm/stablelm_backbone.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+from keras_hub.src.models.backbone import Backbone
+from keras_hub.src.models.stablelm.stablelm_decoder import StableLMTransformerDecoder
+
+def _stablelm_kernel_initializer(stddev=0.02):
+    """Initializer for StableLM kernel weights."""
+    return keras.initializers.RandomNormal(stddev=stddev)
+
+@keras_hub_export("keras_hub.models.StableLMBackbone")
+class StableLMBackbone(Backbone):
+    """
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/stablelm/stablelm_backbone.py:16]: first one line description should be immediately after """

File: keras_hub/src/models/stablelm/stablelm_backbone.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        rope_scaling_factor (float, optional): The scaling factor for RoPE. Defaults to 1.0.
+        layer_norm_epsilon (float, optional): Epsilon for LayerNorm. Defaults to 1e-5.
+        dropout (float, optional): Dropout rate. Defaults to 0.0.
+        tie_word_embeddings (bool, optional): Whether to tie input and output embeddings.
+            Defaults to False.
+        dtype: The dtype to use for computations and weights.
+
+    Example:
+    ```python
+    # Randomly initialized StableLM decoder with custom config
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/stablelm/stablelm_backbone.py:48]: add a from preset example

File: keras_hub/src/models/stablelm/stablelm_backbone.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+                distribution.
+            model_parallel_dim_name: The axis name of the device mesh, where
+                the weights should be partition on.
+            data_parallel_dim_name: The axis name of the device mesh, where
+                the data should be partition on.
+        Return:
+            `keras.distribution.LayoutMap` that contains the sharding spec
+            for all the model weights.
+        """
+        # The weight path and shape of the Llama backbone is like below
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/stablelm/stablelm_backbone.py:226]: remove commented code

**Comment by divyashreepathihalli** [comment]
The numerics is good enough!

File: keras_hub/src/models/stablelm/stablelm_backbone.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    partial rotary position embeddings and LayerNorm with learned bias terms.
+    It includes the embedding lookups and transformer layers.
+
+    The default constructor provides a fully customizable, randomly initialized
+    StableLM model with any number of layers, heads, and embedding dimensions.
+
+    Args:
+        vocabulary_size: int. The size of the token vocabulary.
+        num_layers: int. The number of transformer layers (32 for
+            StableLM-3B4E1T).
```
Comment (abheesht17) [inline keras_hub/src/models/stablelm/stablelm_backbone.py:34]: We probably don't need `(32 for StableLM-2B4E1T)` here (and elsewhere too), since it'll be part of the config anyway. Let's remove.

**Review by abheesht17 — COMMENTED** [review]
Looks good overall, but let's check the numerics and the generate output.

**Comment by mattdangerw** [comment]
@Bond099 let's sync this with the latest changes and make sure to run our format script. I'm not exactly sure why non of our CI is running, but I don't think it ran.

**Comment by abheesht17** [comment]
Let's clean up the PR. Can we fix the following minor things?

- Pull in master.
- I still don't see `tie_weights = False` in the PR. 
- Run formatting, etc.
- Let's wait for all the tests to run.

**Comment by abheesht17** [comment]
Looks like there are conflicts. Please pull in master and resolve conflicts

File: keras_hub/src/samplers/sampler.py [hunk 7]
```diff
@@ -3,7 +3,6 @@
 from keras import random
```
Comment (abheesht17) [inline keras_hub/src/samplers/sampler.py:3]: I'd actually want `sampler.py` to be untouched, because all models use this. We'd want to make this change only for Stable LM. The changes will be less intrusive that way. 

Is there a way you can pass the correct mask right here: https://github.com/keras-team/keras-hub/blob/e836a784ea770b1082e1161caeeea8f35a1d9a43/keras_hub/src/models/stablelm/stablelm_causal_lm.py#L130, https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/samplers/sampler.py#L79?

If you do pass the correct mask, this should evaluate to the correct bool: https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/samplers/sampler.py#L100.

Take a look if this works.

**Comment by sachinprasadhs** [comment]
@Bond099 , Could you please resolve the review comments and update the PR. Thanks!
